### PR TITLE
IGNITE-18558 Fix Meta Storage prefix overflow

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-18558

### CODE REVIEW

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java [hunk 1]
```diff
@@ -676,17 +676,15 @@ public Cursor<Entry> range(ByteArray keyFrom, @Nullable ByteArray keyTo) throws
      * @see ByteArray
      * @see Entry
      */
-    public @NotNull Cursor<Entry> prefixWithAppliedRevision(@NotNull ByteArray keyPrefix) throws NodeStoppingException {
+    public Cursor<Entry> prefixWithAppliedRevision(ByteArray keyPrefix) throws NodeStoppingException {
         if (!busyLock.enterBusy()) {
```
Comment (tkalkirill) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:680]: U can use `org.apache.ignite.internal.util.IgniteUtils#inBusyLock(org.apache.ignite.internal.util.IgniteSpinBusyLock, java.util.function.Supplier<T>)`

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 2]
```diff
@@ -253,19 +254,27 @@ public void onWrite(Iterator<CommandClosure<WriteCommand>> iter) {
...
-                cursors.put(
-                        cursorId,
-                        new CursorMeta(
-                                cursor,
-                                CursorType.RANGE,
-                                rangeCmd.requesterNodeId(),
-                                rangeCmd.batchSize()
-                        )
-                );
+                var meta = new CursorMeta(cursor, CursorType.RANGE, rangeCmd.requesterNodeId(), rangeCmd.batchSize());
```
Comment (tkalkirill) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:261]: I think `metaCursor` would be a better name.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 3]
```diff
@@ -253,19 +254,27 @@ public void onWrite(Iterator<CommandClosure<WriteCommand>> iter) {
...
+
+                IgniteUuid cursorId = prefixCmd.cursorId();
+
+                Cursor<Entry> cursor = prefixCmd.revUpperBound() == -1
+                        ? storage.prefix(prefixCmd.prefix(), prefixCmd.includeTombstones())
+                        : storage.prefix(prefixCmd.prefix(), prefixCmd.revUpperBound(), prefixCmd.includeTombstones());
+
+                var meta = new CursorMeta(cursor, CursorType.RANGE, prefixCmd.requesterNodeId(), prefixCmd.batchSize());
+
+                cursors.put(cursorId, meta);
```
Comment (tkalkirill) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:277]: I think `metaCursor` would be a better name.

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java [hunk 4]
```diff
@@ -359,21 +360,30 @@ public StatementResult invoke(If iif) {
...
-    /** {@inheritDoc} */
     @Override
     public Cursor<Entry> range(byte[] keyFrom, byte[] keyTo, long revUpperBound, boolean includeTombstones) {
         return new RangeCursor(keyFrom, keyTo, revUpperBound, includeTombstones);
     }
 
-    /** {@inheritDoc} */
+    @Override
+    public Cursor<Entry> prefix(byte[] prefix, boolean includeTombstones) {
+        synchronized (mux) {
```
Comment (tkalkirill) [inline modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:377]: Maybe you should add `synchronized` for `prefix`* methods header, for example `public synchronized Cursor<Entry> prefix(...)`?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 1 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:680]: Why should I?

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 4 above)
Comment (sashapolo) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:377]: How would that work? All other methods use a separate object for synchronisation

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:680]: This **suggestion** will reduce the amount of similar code.

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 4 above)
Comment (tkalkirill) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:377]: Write a comment why `mux` is used there.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:680]: The balance of "force" began to shift =)

**Review by tkalkirill — APPROVED** [review]

**Pull request merged**